cli/tracing.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"context"
+	"io"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/uber/jaeger-client-go/config"
+)
+
+// Implements jaeger-client-go.Logger interface
+type traceLogger struct {
+}
+
+// tracerCloser contains a copy of the closer returned by createTracer() which
+// is used by stopTracing().
+var tracerCloser io.Closer
+
+func (t traceLogger) Error(msg string) {
+	kataLog.Error(msg)
+}
+
+func (t traceLogger) Infof(msg string, args ...interface{}) {
+	kataLog.Infof(msg, args...)
+}
+
+func createTracer(name string) (opentracing.Tracer, error) {
```

Comment:
Check if `tracing` is enabled first? Or am I misunderstanding it that `tracing` is checked somewhere else?

Reply:
@jodh-intel Yeah, I think we need to decide who controls the tracing functionality -- kata config or Jaeger config? Right now, each of them has a switch and it looks like only the Jaeger one is in effect.

And do we really need to create these hierarchical spanners? If not, we can put in `app.Before` and `app.After` something like:
```
if (tracing) {
    setup tracer
    start tracer span
}
```
and
```
if (tracing) {
     stop tracer span
     cleanup tracer
}
```
So we have one tracer and one span for each command line invocation. We can pass around `span` so that we can call `SetTag` on it. And it looks much cleaner than having every one to call
```
span := opentracing.StartSpanFromContext
defer span.Finish()
```

Another question about opentracing (sorry to be dumb), do we need to plant trace points on our own? Or is it integrated with logrus somehow? Or can Jaeger insert new trace points dynamically?

Reply:
I agree. But the recommendation is to always pass a `context.Context`, and not to pass the spans directly as that "pollutes" the codebase too much. See:

- https://github.com/yurishkuro/opentracing-tutorial/tree/master/go/lesson02#propagate-the-in-process-context
 (note that that was written by Yuri, the tech lead for Jaeger tracing @ Uber).

Reply:
@jodh-intel Thanks for the presentation. I now understand better how Jaeger works and I agree we need to create span for each function we want to trace. And Jaeger looks more like a function tracing system to me. I have two more questions though:

1. Can we use it to trace any point in the code? If so, we can use it to replace most of the logs.
2. Do users have to install Jaeger (server?) to be able to disable tracing?

Reply:
Hi @bergwolf.

> Can we use it to trace any point in the code?

Yes, assuming the code has access to a `context.Context`.

> Do users have to install Jaeger (server?) to be able to disable tracing?

To disable tracing, they just never uncomment `enable_tracing` in the config file. So if `enable_tracing == false`, the runtime will use the "NOP tracer". But also, if `enable_tracing == true` but no Jaeger agent is running, that will not actually generate any errors either - it will just make the runtime a little slower as the jaeger-client-go code tries (unsucessfully but quietly :) to talk to the Jaeger agent over UDP.

Reply:
... And yes, I'm starting to think that once we have the tracing support working well, we can remove some of the logging. It would be interesting to know how [`zap`](https://github.com/uber-go/zap) performs compared to jaeger since if jaeger is faster (and I'm guessing that it is), we don't need tracing any more in theory. Some users might be concerned about the UDP comms, but that's what syslog uses I believe ;)